src/mixins/yield-erc/ERC20Vault.sol

```diff
@@ -0,0 +1,149 @@
+    /// @param to The address to receive underlying tokens corresponding to the withdrawal.
+    /// @param underlyingAmount The amount of underlying tokens to withdraw.
+    function withdraw(address to, uint256 underlyingAmount) external virtual returns (uint256 shares) {
+        shares = underlyingAmount.fdiv(exchangeRate(), baseUnit);
+
+        // Determine the equivalent amount of shares and burn them.
+        // This will revert if the user does not have enough shares.
+        _burn(msg.sender, shares);
+
+        emit Withdraw(msg.sender, to, underlyingAmount);
+
+        // Withdraw from strategies if needed and transfer.
+        beforeWithdraw(underlyingAmount);
+
+        underlying.safeTransfer(to, underlyingAmount);
+    }
+
+    /// @notice Redeem a specific amount of shares for underlying tokens.
+    /// @param to The address to receive underlying tokens corresponding to the withdrawal.
+    /// @param shareAmount The amount of shares to redeem for underlying tokens.
+    function redeem(address to, uint256 shareAmount) external virtual returns (uint256 underlyingAmount) {
+
+        // Determine the equivalent amount of underlying tokens.
+        underlyingAmount = shareAmount.fmul(exchangeRate(), baseUnit);
+
+        // Burn the provided amount of shares.
+        // This will revert if the user does not have enough shares.
+        _burn(msg.sender, shareAmount);
+
+        emit Withdraw(msg.sender, to, underlyingAmount);
+
+        // Withdraw from strategies if needed and transfer.
+        beforeWithdraw(underlyingAmount);
+
+        underlying.safeTransfer(to, underlyingAmount);
+    }
+
+    function beforeWithdraw(uint256 underlyingAmount) internal virtual {}
+
+    function afterDeposit(uint256 underlyingAmount) internal virtual {}
+
+    /*///////////////////////////////////////////////////////////////
+                        VAULT ACCOUNTING LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Returns a user's Vault balance in underlying tokens.
+    /// @param user The user to get the underlying balance of.
+    /// @return The user's Vault balance in underlying tokens.
+    function balanceOfUnderlying(address user) external view returns (uint256) {
+        return balanceOf[user].fmul(exchangeRate(), baseUnit);
+    }
+
+    /// @notice Returns the amount of underlying tokens an share can be redeemed for.
+    /// @return The amount of underlying tokens an share can be redeemed for.
+    function exchangeRate() public view returns (uint256) {
+        // Get the total supply of shares.
+        uint256 shareSupply = totalSupply;
+
+        // If there are no shares in circulation, return an exchange rate of 1:1.
+        if (shareSupply == 0) return baseUnit;
+
+        // Calculate the exchange rate by dividing the total holdings by the share supply.
+        return totalHoldings().fdiv(shareSupply, baseUnit);
```

Comment:
This may not be a big of an issue, but item 3 in yearn vaults audit may apply.

https://github.com/yearn/yearn-security/blob/master/audits/20210719_ToB_yearn_vaultsv2/ToB_-_Yearn_Vault_v_2_Smart_Contracts_Audit_Report.pdf

First depositor can mint 1 share and then transfer underlying affecting rounding errors depending on base unit?

Reply:
at least in rari we vaults we planned to handle this as uni does in an initialization module, should probably think about how we do this more generally tho, thanks for the reminder!

Reply:
Just passing by, `an share -> a share`. Great reads around here!

Reply:
good find @Elyx0 !

Reply:
updated the comment